File: convexhull.cpp
#include "SDL.h"
#include <bits/stdc++.h>

/*******************************************************
 * TODO:
 * Read through andrews algorithm in book and online
 * Sketch out solution
 * Finish coding
 ******************************************************/






using std::ostream, std::cout, std::ios, std::cin,std::vector;
using std::sort;
//////////////////////////////////////////////////////////////
// globals
//////////////////////////////////////////////////////////////
constexpr size_t NUM_PT = 11;
constexpr int SCALE = 4;
constexpr int WINDIM = 600;

//////////////////////////////////////////////////////////////
// types
//////////////////////////////////////////////////////////////
struct pt {
  float x, y;
  friend ostream& operator<<(ostream& os, const pt& p) {
    os << "x: " << p.x << "\ty:" << p.y;
    return os;
  }
  // comparator
};

// sorting predicate for pt
struct less_than
{
  inline bool operator() (const pt& p1, const pt& p2)
  {
    if (p1.x < p2.x) return true;
    if (p2.x < p1.x) return false;
    // p1=p2 for x, go to y
    if (p1.y < p2.y) return true;
    if (p2.y < p1.y) return false;
    return false;
  }
};

//////////////////////////////////////////////////////////////
// free functions
//////////////////////////////////////////////////////////////

int rand_num(const int lower, const int upper) {
  return ( rand() % (upper - lower + 1)) + lower; 
}

pt create_rand_pt(int first, int last) {
  auto num1 = rand_num(first,last);
  auto num2 = rand_num(first,last);
  return pt{ static_cast<float>(num1), static_cast<float>(num2) };
}

// function to determine if points a--b--c create a left hand turn
bool is_left_turn(pt a, pt b, pt c) {
  // Calculate the angle formed by 3 points.
  // e.g. for a---b---c, it is the angle <(ba,bc)
  auto angle_formed = [](pt a, pt b, pt c) {
    pt ba = pt{a.x-b.x, a.y-b.y}; 
    pt bc = pt{c.x-b.x, c.y-b.y}; 
    float dotprod = ba.x * bc.x + ba.y * bc.y;
    float lenAB = sqrt(ba.x*ba.x + ba.y*ba.y);
    float lenBC = sqrt(bc.x*bc.x + bc.y*bc.y);
    auto ans_radians = acos( (float)dotprod / lenAB / lenBC);
    return ans_radians * 180.0 / M_PI;
  };

  /* For the convex hull Andrews algorithm, need to be able to check
   * if the last 3 points are a left turn or a right turn. If a left
   * turn, throw out the 2nd to last point and start again.
   *
   * create a perpendicular line to the right of the ab.
   * transform ab with the rotation matrix.
   * add b to the transformed coordinates to get d, the point
   * out 90 degrees clockwise to a-b.
   * calculate angle <dbc. If that is less than or equal
   * to 90, it is a righthand turn.*/

  pt ab{b.x - a.x, b.y - a.y};
  // The 90 degrees clockwise rotation matrix is 
  //  0   1
  // -1   0
  //  Obtain d by rotating ab 90 degrees clockwise and adding
  //  it to point b.
  pt d{ab.y + b.x, -ab.x + b.y};
  // If the angle formed by dbc is <= 90, its a right hand turn or
  // abc is a straight line. 
  // Convex hulls are made from all right hand turns.
  // If the angle formed is > 90, it is a left turn.
  return angle_formed(d,b,c) > 90.f;
} 



vector<pt> solve_convex_hull(vector<pt> pts) {
  vector<pt> soln(NUM_PT);
  assert(soln.size() == NUM_PT);
  soln.resize(0);
  // step 1, sort by x then y
  // add first 3 points
  // Loop:
  // if left_turn == 1, remove 2nd to last point
  // add a point
  // go until no more pts
  sort(pts.begin(), pts.end(), less_than());
  //  print out sorted pts array
  cout << "SORTED POINTS:\n";
  for (int i = 0; i < NUM_PT; i++)
    cout << pts[i] << "\n";
  size_t nextpt = 0;
  // add first three points to soln
  soln.push_back(pts[0]);
  soln.push_back(pts[1]);
  nextpt = 2;
  while(nextpt != pts.size()+1) {
    soln.push_back(pts[nextpt]);
    auto sz = soln.size();
    if (is_left_turn(soln[sz-3],soln[sz-2],soln[sz-1]))
      soln.erase(soln.end()-2);
    nextpt++;
  }
  // do one final check in case the last point added was bad
  auto sz = soln.size();
  if (is_left_turn(soln[sz-3],soln[sz-2],soln[sz-1]))
    soln.erase(soln.end()-2);

  //  print out soln array
  cout << "SOLUTION POINTS (" << soln.size() << "%ld):\n";
  for (int i = 0; i < soln.size(); i++)
    cout << soln[i] << "\n";

  return soln;

}

//////////////////////////////////////////////////////////////
// main function
//////////////////////////////////////////////////////////////
int main(int argc, char* argv[])
{
  // random points 
  srand(time(NULL));
  vector<pt> pts(NUM_PT);

  for (int i = 0; i < NUM_PT; i++) {
    pts[i] = create_rand_pt(80/SCALE,(WINDIM-80)/SCALE);
  }

  vector<pt> soln = solve_convex_hull(pts);



  // create window and draw solution

  if (SDL_Init(SDL_INIT_VIDEO) == 0) {
    SDL_Window* window = NULL;
    SDL_Renderer* renderer = NULL;

    if (SDL_CreateWindowAndRenderer(WINDIM, WINDIM, 0, &window, &renderer) == 0) {
      SDL_bool done = SDL_FALSE;

      SDL_RenderSetScale(renderer, SCALE, SCALE);

      while (!done) {
        SDL_Event event;

        SDL_SetRenderDrawColor(renderer, 0, 0, 0, SDL_ALPHA_OPAQUE);
        SDL_RenderClear(renderer);

        // draw the solution
        SDL_SetRenderDrawColor(renderer, 100, 5, 200, SDL_ALPHA_OPAQUE);
        auto num_lines = soln.size() - 1;
        // a---b---c----d
        for (int i = 0; i < num_lines -1; i++) {
          SDL_RenderDrawLine(renderer, soln[i].x, WINDIM/SCALE-soln[i].y, \
              soln[i+1].x, WINDIM/SCALE-soln[i+1].y);
        }
       
        // draw the points
        SDL_SetRenderDrawColor(renderer, 255, 255, 255, SDL_ALPHA_OPAQUE);
        for (int i = 0; i < NUM_PT; i++) 
          SDL_RenderDrawPoint(renderer, pts[i].x, WINDIM/SCALE-pts[i].y);

        SDL_RenderPresent(renderer);

        while (SDL_PollEvent(&event)) {
          if (event.type == SDL_QUIT) {
            done = SDL_TRUE;
          }
        }
      }
    }

    if (renderer) {
      SDL_DestroyRenderer(renderer);
    }
    if (window) {
      SDL_DestroyWindow(window);
    }
  }
  SDL_Quit();
  return 0;
}

File: convex-hull.r
library(magrittr)
library(ggplot2)
# clear objects
rm(list = ls())
pts = list()
length(pts) = 60 

rnd = function() {
  # random number in specified range
  sample(25:125, 1)
}

generate_rand_2vecs = function(pts) {
  for (i in 1:length(pts)) {
    pts[[i]] = c(rnd(), rnd()) %>% matrix
  }
  pts
}
pts = generate_rand_2vecs(pts)
df = data.frame(data = pts)
df = df %>% t
rownames(df) = 1:length(pts)
colnames(df) = c("x", "y")
rm(pts)
print(df)
df = data.frame(df)


selectV2 = function(df, i) {
  df[i, ] %>% t %>% as.vector(.,mode="double")
}
norm_vec <- function(x) { sqrt(sum(x^2)) }
deg2rad = function(x) {
  x * pi / 180.0 
}
rad2deg = function(x) {
  x * 180.0 / pi 
}
rotMat = function(ang) {
  # create clockwise rotation matrix given an angle in degrees
  # 1st convert degrees to radians
  ang = deg2rad(ang)
  # since the below is a counterclockwise rotation matrix,
  # convert to angle from ccw to cw
  ang = 2 * pi - ang
  res = matrix(c(cos(ang), -sin(ang), sin(ang),
                 cos(ang)), ncol = 2, byrow = TRUE)
  return(res)
}
sort.less.than = function(df) {
  #return(df[order(df$x,df$y),])
  df = df[order(df$y,decreasing = TRUE),]
  df = df[order(df$x),]
  return(df)
  #return(df[order(df$x,df$y),])
}
sort.greater.than = function(pts) {
  df = df[order(df$y, decreasing = FALSE),]
  df = df[order(df$x, decreasing = TRUE),]
  return(df)
  #return(df[order(df$x,df$y,decreasing = TRUE),])
}
angle.formed = function(a,b,c) {
  # angle in degrees formed by three 2-vectors
  stopifnot(length(a) == 2 && typeof(a) == "double")   
  stopifnot(length(b) == 2 && typeof(b) == "double")   
  stopifnot(length(c) == 2 && typeof(c) == "double")   
  ba = (a - b) %>% as.matrix
  bc = (c - b) %>% as.matrix
  dp = crossprod(ba,bc) 
  angRad = acos(dp / norm_vec(ba) / norm_vec(bc))
  return(angRad %>% rad2deg)
}
is.left.turn = function(a,b,c) {
  ab = b - a
  rotated = rotMat(90) %*% ab
  d = rotated + b
  return(angle.formed(d,b,c) > 90.0)
}
add.pt = function(df, soln, currIndex) {
  rbind(soln, selectV2(df,currIndex)) 
}

solve.convex.hull = function(upper=TRUE) {
  soln = matrix(ncol=2,nrow=0)
  # sort by decreasing to create the top of the hull
  if (upper==TRUE){
  df = sort.less.than(df)
  } else {
  df = sort.greater.than(df)
  }
  # add first two points
  for(i in 1:2) { soln = add.pt(df, soln,i) }
  # keep track of the next point from the sorted set to add to
  # the solution.
  curridx = 3
  
  
  # solve top of hull
  # add a point
  # check if left turn
  # if left turn, remove 2nd to last point in soln
  last.pt.added.to.soln = c(-1,-1)
  while (curridx <= length(df[, 1])) {
    soln = add.pt(df, soln, curridx)
    last.pt.added.to.soln = soln[length(soln[,1]),]
    # Remove 2nd to last point if a left hand turn is 
    # formed by the last 3 points in the solution.
    # Continue removing until either they form a right hand turn
    # or there are only 2 points in the solution.
    while (1) {
      lenSoln = length(soln[, 1])
      if (lenSoln < 3) break;
      removeFlag = is.left.turn(soln[lenSoln - 2, ],
                                soln[lenSoln - 1, ],
                                soln[lenSoln - 0, ])
      if (!removeFlag) break;
        # remove 2nd to last point
        soln = soln[-(lenSoln - 1), ]
    }
    curridx = curridx + 1
  }
  colnames(soln) = c("x", "y")
  if (upper==FALSE) {
  print(last.pt.added.to.soln)
  }
  data.frame(soln)
}



test.convex.hull = function() {
  soln = matrix(ncol=2,nrow=0)
  a = c(1,2); b = c(2,3.5); c = c(1.5,4.5);
  soln = rbind(soln,a)
  soln = rbind(soln,b)
  soln = rbind(soln,c)
  angle.formed(soln[1,],soln[2,],soln[3,])
  is.left.turn(soln[1,],soln[2,],soln[3,])
  
  soln
  lenSoln = length(soln[,1])
  data.frame(soln)
}

#test.convex.hull()
soln.upper = solve.convex.hull(upper=TRUE)
soln.lower = solve.convex.hull(upper=FALSE)

cols = c("blue" = "blue")
# plotting the result
ggplot() +
  layer(
    data = df,
    mapping = aes(x = x, y = y),
    stat = "identity",
    position = "identity",
    geom = "point"
  ) +
  xlim(c(0, 150)) +
  ylim(c(0, 150)) + 
  layer(
    data = soln.upper, 
    stat = "identity",
    position = "identity",
    geom = "path",
    mapping = aes(x = x, y = y),
  ) + 
  layer(
    data = soln.lower, 
    mapping = aes(x = x, y = y),
    stat = "identity",
    position = "identity",
    geom = c("path")
  ) 
